politeiawww/user/mysql/mysql.go

```diff
@@ -0,0 +1,1081 @@
+	defer cancel()
+
+	// Start transaction.
+	opts := &sql.TxOptions{
+		Isolation: sql.LevelDefault,
+	}
+	tx, err := m.userDB.BeginTx(ctx, opts)
+	if err != nil {
+		return fmt.Errorf("begin tx: %v", err)
+	}
+	defer tx.Rollback()
+
+	err = setPaywallAddressIndex(ctx, tx, index)
+	if err != nil {
+		return err
+	}
+
+	// Commit transaction.
+	if err := tx.Commit(); err != nil {
+		if err2 := tx.Rollback(); err2 != nil {
+			// We're in trouble!
+			panic(fmt.Errorf("rollback tx failed: commit:'%v' rollback:'%v'",
+				err, err2))
+		}
+		return fmt.Errorf("commit tx: %v", err)
+	}
+
+	return nil
+}
+
+// RotateKeys rotates the existing database encryption key with the given new
+// key.
+func (m *mysql) RotateKeys(newKeyPath string) error {
+	log.Tracef("RotateKeys: %v", newKeyPath)
+
+	if m.isShutdown() {
+		return user.ErrShutdown
+	}
+
+	// Load and validate new encryption key.
+	newKey, err := util.LoadEncryptionKey(log, newKeyPath)
+	if err != nil {
+		return fmt.Errorf("load encryption key '%v': %v",
+			newKeyPath, err)
+	}
+
+	if bytes.Equal(newKey[:], m.encryptionKey[:]) {
+		return fmt.Errorf("keys are the same")
+	}
+
+	log.Infof("Rotating encryption keys")
+
+	ctx, cancel := ctxWithTimeout()
+	defer cancel()
+
+	m.Lock()
+	defer m.Unlock()
+
+	// Rotate keys using a transaction.
+	opts := &sql.TxOptions{
+		Isolation: sql.LevelDefault,
+	}
+	tx, err := m.userDB.BeginTx(ctx, opts)
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+	err = rotateKeys(ctx, tx, m.encryptionKey, newKey)
+	if err != nil {
+		return err
+	}
+
+	// Commit transaction.
+	if err := tx.Commit(); err != nil {
+		if err2 := tx.Rollback(); err2 != nil {
+			// We're in trouble!
+			panic(fmt.Errorf("rollback tx failed: commit:'%v' rollback:'%v'",
+				err, err2))
+		}
+		return fmt.Errorf("commit tx: %v", err)
+	}
+
+	// Update context.
+	m.encryptionKey = newKey
+
+	return nil
+}
+
+// PluginExec executes a plugin command.
+func (m *mysql) PluginExec(pc user.PluginCommand) (*user.PluginCommandReply, error) {
+	log.Tracef("PluginExec: %v %v", pc.ID, pc.Command)
+
+	if m.isShutdown() {
+		return nil, user.ErrShutdown
+	}
+
+	var payload string
+	var err error
+	switch pc.ID {
+	case user.CMSPluginID:
+	default:
+		return nil, user.ErrInvalidPlugin
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return &user.PluginCommandReply{
+		ID:      pc.ID,
+		Command: pc.Command,
+		Payload: payload,
+	}, nil
+}
+
+// Close shuts down the database.  All interface functions must return with
+// errShutdown if the backend is shutting down.
+func (m *mysql) Close() error {
+	log.Tracef("Close")
+
+	m.Lock()
+	defer m.Unlock()
+
+	// Zero out encryption key.
+	util.Zero(m.encryptionKey[:])
+	m.encryptionKey = nil
+
+	m.shutdown = true
+	return m.userDB.Close()
+}
+
+// New connects to a mysql instance using the given connection params,
+// and returns pointer to the created mysql struct.
+func New(host, password, network, encryptionKey string) (*mysql, error) {
```

Comment:
This is hanging on startup for me.  There's a bug somewhere.  I can't test it any further until this is fixed.

```
2021-06-13 15:52:30.154 [INF] PWWW: Network : testnet3
2021-06-13 15:52:30.154 [INF] PWWW: Home dir: /home/luke/.politeiawww
2021-06-13 15:52:30.154 [INF] PWWW: Load CSRF key
2021-06-13 15:52:30.338 [INF] PWWW: User database: mysql
2021-06-13 15:52:30.338 [INF] USER: MySQL host: politeiawww:[password]@tcp(localhost:26257)/users_testnet3
```

Reply:
Oh, looks like you are using the cockrochroach host in the example above, just try to run with `:3306`.

Will be added to README, abd maybe it shouldn't just hang when connection can not be eatablished.